File: kube/src/main/java/com/jacksonrakena/infrastructure/apps/GradekeeperServer.kt
package com.jacksonrakena.infrastructure.apps

import com.jacksonrakena.infrastructure.util.applyCommonConfiguration
import org.cdk8s.Chart
import org.cdk8s.ChartProps
import org.cdk8s.plus28.ContainerPort
import org.cdk8s.plus28.ContainerProps
import org.cdk8s.plus28.Deployment
import org.cdk8s.plus28.DeploymentExposeViaServiceOptions
import org.cdk8s.plus28.DeploymentProps
import org.cdk8s.plus28.DeploymentStrategy
import org.cdk8s.plus28.DeploymentStrategyRollingUpdateOptions
import org.cdk8s.plus28.DockerConfigSecret
import org.cdk8s.plus28.EnvFrom
import org.cdk8s.plus28.EnvValue
import org.cdk8s.plus28.IConfigMap
import org.cdk8s.plus28.PercentOrAbsolute
import org.cdk8s.plus28.Service
import org.cdk8s.plus28.ServicePort
import software.constructs.Construct

class GradekeeperServer(
    scope: Construct,
    id: String,
    configMap: IConfigMap,
    postgresService: Service,
    registrySecret: DockerConfigSecret,
    props: ChartProps? = null,
) : Chart(scope, id, props) {
    val deployment = Deployment(
        this,
        "deployment",
        DeploymentProps.builder()
            .replicas(2)
            .strategy(
                DeploymentStrategy.rollingUpdate(
                    DeploymentStrategyRollingUpdateOptions.builder().maxUnavailable(
                        PercentOrAbsolute.absolute(1)
                    ).build()
                )
            )
            .containers(
                listOf(
                    ContainerProps.builder()
                        .name("server")
                        .image("ghcr.io/jacksonrakena/gradekeeper-server:latest")
                        .applyCommonConfiguration()
                        .ports(
                            listOf(
                                ContainerPort.builder().number(3000).build()
                            )
                        )
                        .envFrom(
                            listOf(
                                EnvFrom(configMap)
                            )
                        )
                        .envVariables(
                            mapOf(
                                "DATABASE_URL" to EnvValue.fromValue("postgresql://gradekeeper:gradekeeper@${postgresService.name}/gradekeeper")
                            )
                        )
                        .build()
                )
            )
            .dockerRegistryAuth(registrySecret)
            .build()
    )

    val service = deployment.exposeViaService(
        DeploymentExposeViaServiceOptions.builder()
            .ports(
                listOf(
                    ServicePort.builder().port(80).targetPort(3000).build()
                )
            ).build()
    )
}

File: src/apps/gradekeeper-server.ts
import { Construct } from "constructs";
import { Chart, ChartProps, Duration } from "cdk8s";
import * as kplus from "cdk8s-plus-28";
import { applyCommonConfiguration } from "../util/container-utils";

export class GradekeeperServer extends Chart {
  public readonly service: kplus.Service;

  constructor(
    scope: Construct,
    id: string,
    configMap: kplus.IConfigMap,
    databaseServiceReadWrite: string,
    databaseSecret: kplus.ISecret,
    registrySecret: kplus.DockerConfigSecret,
    props?: ChartProps,
  ) {
    super(scope, id, props);

    const deployment = new kplus.Deployment(this, "deployment", {
      replicas: 2,
      strategy: kplus.DeploymentStrategy.rollingUpdate({
        maxUnavailable: kplus.PercentOrAbsolute.absolute(1),
      }),
      containers: [
        applyCommonConfiguration({
          name: "server",
          image: "ghcr.io/jacksonrakena/gradekeeper-server:latest",
          ports: [{ number: 3000 }],
          envFrom: [new kplus.EnvFrom(configMap)],
          envVariables: {
            DATABASE_NAME: kplus.EnvValue.fromValue("gradekeeper"),
            DATABASE_HOST: kplus.EnvValue.fromValue(databaseServiceReadWrite),
            DATABASE_PASSWORD: kplus.EnvValue.fromSecretValue({
              key: "password",
              secret: databaseSecret,
            }),
            DATABASE_USERNAME: kplus.EnvValue.fromSecretValue({
              key: "username",
              secret: databaseSecret,
            }),
          },
          liveness: kplus.Probe.fromHttpGet("/health", {
            initialDelaySeconds: Duration.seconds(10),
            port: 3000,
          }),
        }),
      ],
      dockerRegistryAuth: registrySecret,
    });

    this.service = deployment.exposeViaService({
      ports: [{ port: 80, targetPort: 3000 }],
    });
  }
}
